libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/util/CacheResultUtils.kt

```diff
@@ -51,6 +53,106 @@ internal object CacheResultUtils {
+            trafficProvider
+        )
+        override fun equals(other: Any?): Boolean {
+            if (this === other) return true
+            if (javaClass != other?.javaClass) return false
+
+            other as CacheResultKeyRouteTraffic<*>
+
+            if (route.id != other.route.id) return false
+            if (trafficProvider != trafficProvider) return false
+            if (
+                route.directionsRoute.legs()?.size !=
+                other.route.directionsRoute.legs()?.size
+            ) {
+                return false
+            }
+            if (!trafficIsEqual(route, other.route)) return false
+
+            return true
+        }
+
+        override fun hashCode(): Int {
+            var result = route.id.hashCode()
+            result = 31 * result + trafficProvider.hashCode()
+            route.directionsRoute.legs()?.forEach { routeLeg ->
+                result = 31 * result + routeLeg.annotation()?.congestion().hashCode()
+                result = 31 * result + routeLeg.annotation()?.congestionNumeric().hashCode()
+            }
+            return result
+        }
+
+        private fun trafficIsEqual(route: NavigationRoute, other: NavigationRoute): Boolean {
+            val hasNumericTraffic: Boolean = route.directionsRoute.routeOptions()?.annotationsList()
+                ?.contains(DirectionsCriteria.ANNOTATION_CONGESTION_NUMERIC) ?: false
+            val sameTrafficType = hasNumericTraffic ==
+                other.directionsRoute.routeOptions()?.annotationsList()
+                    ?.contains(DirectionsCriteria.ANNOTATION_CONGESTION_NUMERIC)
+            when (sameTrafficType) {
+                false -> return false
+                true -> {
+                    route.directionsRoute.legs()?.forEachIndexed { index, routeLeg ->
+                        val trafficCongestionIsEqual = when (hasNumericTraffic) {
+                            false -> {
+                                if (
+                                    routeLeg.annotation()?.congestion()?.size !=
+                                    other.directionsRoute
+                                        .legs()?.get(index)?.annotation()?.congestion()?.size
+                                ) {
+                                    false
+                                } else {
+                                    listElementsAreEqual(
+                                        routeLeg.annotation()?.congestion() ?: listOf(),
+                                        other.directionsRoute
+                                            .legs()
+                                            ?.get(index)
+                                            ?.annotation()
+                                            ?.congestion() ?: listOf()
+                                    ) { string1: String, string2: String -> string1 === string2 }
+                                }
+                            }
+                            true -> {
```

Comment:
Can you generalize 2 branches here (when hasNumericTraffic is true or false)? It only differs by congestion/congestionNumeric invocation and the comparator.

Reply:
If one route has numeric traffic and the other doesn't the routes aren't the same in the context of this use case.

Reply:
I meant that when `sameTrafficType` is true, the inner when branches differ only by congestion extractor (it's either `congestion` or `congestionNumeric`).
Nothing critical, but it could reduce the risk of forgetting to add something to the other branch when the logic will be modified in the future.